### PR TITLE
DDF-3477 Add attribute sorting to SortedQueryMonitor

### PR DESCRIPTION
**Pull request by troymohl**
#### What does this PR do?
Adds attribute sorting to SortedQueryMonitor

#### Who is reviewing it? 
@adimka 
@mackncheesiest  

#### Select relevant component teams: 
https://github.com/orgs/codice/teams

#### Choose 2 committers to review/merge the PR. 
@bdeining
@millerw8
@pklinef 

#### Any background context you want to provide?
Currently sorting is only done by relevance, temporal, or distance. This adds the ability to sort by other attributes when aggregating results from multiple sources.

#### What are the relevant tickets?
[DDF-3477](https://codice.atlassian.net/browse/DDF-3477)

#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
For the general case, merging is gated on the following:

* Review by at least two contributors
* Review by at least two committers
* Successful CI
* Successful static analysis (SonarQube/Coverity/etc.)
* Heroing of change (details vary by ticket)

Certain exceptions may occur. For example, if there is a problem with the CI environment then a manually performed full build and static analysis check may be substituted _at the discretion of the repository maintainers_. Additionally, repository metadata such as this `PULL_REQUEST_TEMPLATE`, the repository `README`, and other non-functional items may be merged without requiring CI and static analysis.

Finally, an abbreviated review process may be in place for forward-port tickets immediately following branching and prior to branch divergence. In that case, approvals, CI, static analysis, and heroing that occur against the PR against the _release branch_ can be considered sufficient for reviewing the PR against the _master branch_.

Once the branches diverge, the process will return to normal.

Insofar as divergence is not a function of time but of change to the parent branch, the repository maintainers will determine when abbreviated review ends.

When abbreviated review is in effect, a note to that effect will be at the top of this template. Following is an example note:
```
##### ABBREVIATED REVIEW BETWEEN 2.12.X AND MASTER IS IN EFFECT
____

```

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/BasicAttributeComparator.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      } else {
+        Serializable attrValueA = contentAttrA.getValue();
+        Serializable attrValueB = contentAttrB.getValue();
+
+        if (attrValueA == null && attrValueB != null) {
+          return -1;
+        } else if (attrValueA != null && attrValueB == null) {
+          return 1;
+        } else {
+          int compareValue = ComparatorUtils.NATURAL_COMPARATOR.compare(attrValueA, attrValueB);
```
Comment (paouelle) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/BasicAttributeComparator.java]: ❗️ Will throw NPE if both values are null

File: catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/util/impl/BasicAttributeComparatorTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        basicAttributeComparator.compare(getTestResult(2.123), getTestResult(1.123));
+    assertThat(compareResult, is(1));
+
+    compareResult = basicAttributeComparator.compare(getTestResult(1.123), getTestResult(2.123));
+    assertThat(compareResult, is(-1));
+
+    compareResult = basicAttributeComparator.compare(getTestResult(1.123), getTestResult(1.123));
+    assertThat(compareResult, is(0));
+  }
+
```
Comment (paouelle) [inline catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/util/impl/BasicAttributeComparatorTest.java]: ❗️ Missing tests with null value on first, on second, on both

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java [hunk 3]
```diff
@@ -128,10 +129,10 @@ public void run() {
           comparator = new DistanceResultComparator(sortOrder);
         } else if (Result.RELEVANCE.equals(sortType)) {
           comparator = new RelevanceResultComparator(sortOrder);
+        } else {
+          comparator = new BasicAttributeComparator(sortingProp.getPropertyName(), sortOrder);
```
Comment (paouelle) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Suggestion, the above class could be simplified with:
```
comparator = ComparatorUtils.transformedComparator(
      ComparatorUtils
          .nullLowComparator(
              (sortOrder == sortOrder.ASCENDING) ? ComparatorUtils.NATURAL_COMPARATOR
                  : ComparatorUtils.reversedComparator(ComparatorUtils.NATURAL_COMPARATOR)),
      new Transformer<Result, Serializable>() {
        @Override
        public Serializable transform(Result r) {
          if (r == null) {
            return null;
          }
          final Attribute a = r.getMetacard().getAttribute(sortingProp.getPropertyName());

          return (a != null) ? a.getValue() : null;
        }
      });```

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/BasicAttributeComparator.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.util.impl;
+
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.Result;
+import java.io.Serializable;
+import java.util.Comparator;
+import org.apache.commons.collections.ComparatorUtils;
```
Comment (paouelle) [inline catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/BasicAttributeComparator.java]: ✏️  Should upgrade to collections4

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java (hunk 3 above)
Comment (bdeining) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Curious : What is the benefit to condensing the code in this way?

Comment (bdeining) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Is this a ✏️  comment?

Comment (paouelle) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Less code to maintain since you are re-using existing code in Java. The above already had one bug. This code simply takes advantage of the Comparator.nullsFirst() and the Comparator.comparing() methods

Comment (rzwiefel) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: There's no logic to read through , which makes it simpler to understand and maintain.

Comment (troymohl) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Updating the code to match @paouelle suggestion and removing BasicAttributeComparator.java

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java [hunk 5]
```diff
@@ -128,10 +129,16 @@ public void run() {
           comparator = new DistanceResultComparator(sortOrder);
         } else if (Result.RELEVANCE.equals(sortType)) {
           comparator = new RelevanceResultComparator(sortOrder);
+        } else {
+          comparator =
+              Comparator.comparing(
+                  r -> getAttributeValue((Result) r, sortingProp.getPropertyName()),
+                  Comparator.<Comparable>nullsFirst(
```
Comment (millerw8) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Do we want nulls first regardless of direction?  Maybe nulls last if DESCENDING?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java [hunk 6]
```diff
@@ -128,10 +129,16 @@ public void run() {
           comparator = new DistanceResultComparator(sortOrder);
         } else if (Result.RELEVANCE.equals(sortType)) {
           comparator = new RelevanceResultComparator(sortOrder);
+        } else {
+          comparator =
+              Comparator.comparing(
+                  r -> getAttributeValue((Result) r, sortingProp.getPropertyName()),
```
Comment (millerw8) [inline catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Do date properties come back as date objects?  What if the user tries to sort on created date?

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java (hunk 5 above)
Comment (troymohl) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: Will update code to do this.

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java (hunk 6 above)
Comment (troymohl) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: This will work as `Date` implements `Comparable` https://docs.oracle.com/javase/8/docs/api/java/util/Date.html

File: catalog/core/catalog-core-standardframework/pom.xml [hunk 7]
```diff
@@ -533,12 +533,12 @@
                                         <limit>
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.73</minimum>
+                                            <minimum>0.75</minimum>
```
Comment (paouelle) [inline catalog/core/catalog-core-standardframework/pom.xml]: 👍

File: catalog/core/catalog-core-standardframework/pom.xml [hunk 8]
```diff
@@ -533,12 +533,12 @@
...
                                         </limit>
                                         <limit>
                                             <counter>BRANCH</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.63</minimum>
+                                            <minimum>0.64</minimum>
```
Comment (paouelle) [inline catalog/core/catalog-core-standardframework/pom.xml]: 👍

File: catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java [hunk 9]
```diff
@@ -215,6 +228,106 @@ public void interruptThirdFuture() throws Exception {
...
+
+    SortBy ascendingSort = mock(SortBy.class);
+    when(ascendingSort.getSortOrder()).thenReturn(SortOrder.ASCENDING);
+    when(ascendingSort.getPropertyName()).thenReturn(propertyName);
+
+    SortBy descendingSort = mock(SortBy.class);
+    when(descendingSort.getSortOrder()).thenReturn(SortOrder.DESCENDING);
+    when(descendingSort.getPropertyName()).thenReturn(propertyName);
+
+    testSorting(new String[] {null, "a"}, new String[] {null, "a"}, ascendingSort);
```
Comment (paouelle) [inline catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java]: ✏️  Should we split these 10 unit test cases in 10 separate unit test cases?

File: catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java [hunk 10]
```diff
@@ -215,6 +228,106 @@ public void interruptThirdFuture() throws Exception {
             NullPointerException.class, InterruptedException.class, InterruptedException.class);
   }
 
+  @Test
+  public void testSorts() throws Exception {
```
Comment (paouelle) [inline catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java]: ✏️  Should add tests with null attribute found in the result and with non-`Comparable` values.

File: catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java (hunk 6 above)
Comment (millerw8) [reply catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/cache/solr/impl/SortedQueryMonitor.java]: cool, just wasn't sure they were returned as Date objects

File: catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java (hunk 9 above)
Comment (troymohl) [reply catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java]: Fixed.

File: catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java (hunk 10 above)
Comment (troymohl) [reply catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/SortedQueryMonitorTest.java]: Fixed

**Comment by troymohl** [comment]
build now

**Comment by glenhein** [comment]
I built and installed two instances of DDF. The first one federated via CSW to the second instance. I ingested different content into each instance. I then did a CSW query against the first instance. I had the CSW query sort by title, both ASC and DESC. The results came back in the sorted order as expected.
